apollo-livedata-support/src/main/java/com/apollographql/apollo/livedata/LiveDataApollo.kt

```diff
@@ -0,0 +1,241 @@
+                    watcher.enqueueAndWatch(object : ApolloCall.Callback<T>() {
+                        override fun onResponse(response: Response<T>) {
+                            postValue(ApolloLiveDataResponse.of { response })
+                        }
+
+                        override fun onFailure(e: ApolloException) {
+                            postValue(ApolloLiveDataResponse.Failure.Exception(e))
+                        }
+                    })
+                }
+            }
+        }
+    }
+
+    /**
+     * Converts an [ApolloCall] to a [LiveData]. The number of emissions this LiveData will have is based
+     * on the [com.apollographql.apollo.fetcher.ResponseFetcher] used with the call.
+     *
+     * @param call the ApolloCall to convert.
+     * @param T    the value type.
+     * @return the converted LiveData.
+     * @throws NullPointerException if originalCall == null
+     */
+    fun <T> from(@NonNull call: ApolloCall<T>): LiveData<ApolloLiveDataResponse<T>> {
+        checkNotNull(call, "call == null")
+        return object : LiveData<ApolloLiveDataResponse<T>>() {
+            var started = AtomicBoolean(false)
+
+            @MainThread
+            override fun setValue(value: ApolloLiveDataResponse<T>?) {
+                started.set(true)
+                super.setValue(value)
+            }
+
+            override fun postValue(value: ApolloLiveDataResponse<T>?) {
+                started.set(true)
+                super.postValue(value)
+            }
+
+            override fun onActive() {
+                super.onActive()
+                if (started.compareAndSet(false, true)) {
+                    call.enqueue(object : ApolloCall.Callback<T>() {
+                        override fun onResponse(response: Response<T>) {
+                            postValue(ApolloLiveDataResponse.of { response })
+                        }
+
+                        override fun onFailure(e: ApolloException) {
+                            postValue(ApolloLiveDataResponse.Failure.Exception(e))
+                        }
+
+                        override fun onStatusEvent(event: ApolloCall.StatusEvent) {
+                            if (event == ApolloCall.StatusEvent.COMPLETED) {
+                                postValue(ApolloLiveDataResponse.Complete())
+                            }
+                        }
+                    })
+                }
+            }
+        }
+    }
+
+    /**
+     * Converts an [ApolloPrefetch] to a LiveData.
+     *
+     * @param prefetch the ApolloPrefetch to convert.
+     * @return the converted LiveData.
+     * @throws NullPointerException if prefetch == null
+     */
+    fun <T> from(@NonNull prefetch: ApolloPrefetch): LiveData<ApolloLiveDataResponse<T>> {
+        checkNotNull(prefetch, "prefetch == null")
+        return object : LiveData<ApolloLiveDataResponse<T>>() {
+            var started = AtomicBoolean(false)
```

Comment:
i think i'm confused w/ the started flag (sorry idk about Apollo).
Is there a chance that setValue mgiht be called outside this class (notice, it is protected so someone holding an instance of LiveData cannot call it).

Seems like you only need a normal boolean that gets checked in onActive. You would not need an atomic because onActive is called only on the main thread.

Reply:
Hi, appreciate your detail review!
The started `AtomicBoolean` flag is removed.